### ISSUE TITLE
"Generalizing the types of these mutually recursive definitions requires an incompatible set of universes"

### ISSUE THREAD

**Issue by mtzguido** [FStarLang/FStar#1480]
This is a very confusing error
```
let rec f (x:nat) : nat =
    if x > 0
    then f (x-1)
    else 42

and g y = ()
```
```
A.fst(3,0-8,12): (Error 89) Generalizing the types of these mutually recursive definitions requires an incompatible set of universes for f and g
```
It even happens with `--lax` and I think with `--MLish` too. The error is coming from the uvar in `y`'s type: if one annotates it, then the error is gone.
1. Can we make the error better?
2. Should `--lax`/`--MLish` ignore everything universe-related?

**Comment by nikswamy** [comment]
Here's another failing example where the manner in which we type mutually recursive universe polymorphism is quite limited. 

Thanks to @tahina-pro for this example

```
let rec x (t: Type u#a) (l: list t) : Tot (Type u#(a + 1)) =
  match l with
  | [] → Type u#a
  | z :: q → y t q 
and y (t: Type u#a) (l: list t) : Tot (Type u#(a + 1)) =
  match l with
  | [] → Type u#a
  | z :: q → x t q
```

**Comment by catalin-hritcu** [comment]
Found another example that's similar to Nik's, but maybe even a bit simpler:
```fstar
let rec flatten' (#a:Type u#a) (t:tree a) : list a =
  match t with
  | Leaf v   -> [v]
  | Branch l -> flatten_l u#a l
and flatten_l (#a:Type u#a) (l:list (tree a)) : list a =
  match l with
  | []    -> []
  | hd::tl -> FStar.List.append (flatten' hd) (flatten_l u#a tl)
```
getting the following error for the `l` argument of `flatten_l` (with `--print_universes`):
```
Expected expression of type "l:
list u#uu___55510 (tree u#uu___55510 (*?u1333*) _)
  {LexCons u#_ u#0 l (LexTop u#0) << LexCons u#_ u#0 t (LexTop u#0)}"; got expression "l" of type "list u#uu___55509 (tree u#uu___55509 a)"
```
Here's a complete file: https://gist.github.com/catalin-hritcu/9e0d391791a833bb16c90701b3f9e8df

Guido said it does look like a bug, and thinks F* is trying to generalize both universe variables independently. Here are some workarounds Guido found:
- using any fixed universe level works
- or using doing the letrec internally:
```fstar
let packed_flatten (#a:Type u#aa) (t:tree a) : list a =
  let rec flatten' (#a:Type u#aa) (t:tree a) : list a =
    match t with
    | Leaf v   -> [v]
    | Branch l -> 
      flatten_l  l
  and flatten_l (#a:Type u#aa) (l:list (tree a)) : list a =
    match l with
    | []    -> []
    | hd::tl -> FStar.List.append (flatten' hd) (flatten_l tl)
  in 
  flatten' #a t
```

**Comment by nikswamy** [comment]
The original example is still rejected: Generalizing the universe of `g` without generalizing `f` would give `f` the type `<u>. nat -> nat` and leave no way to infer the instantiation of `u`.

The other examples are accepted.

**Comment by nikswamy** [comment]
Fixed since #2408
